Specify hooks formatting (#108)
php-fig/per-coding-style: spec.md

}
```

> ### 4.4 Property Hooks

Reviewer: IMO we should add this to the end of 4 rather than displacing everything else in 4. Ideally one could refer to `PER-CS 4.5` and expect it to remain the same across _at least_ minor versions.

Reviewer 2: Agree.

Object properties may also include hooks, which have a number of syntactic options.

If there is only one hook implementation, and that hook implementation itself uses
the short-hook syntax, then the hook declaration MAY be listed entirely inline.  The hook
name MUST be separated from the opening brace and the arrow operator by a single

> space, and the semicolon ending of the hook MUST be separated from the cosing brace

Reviewer 3: ```suggestion
space, and the semicolon ending of the hook MUST be separated from the closing brace
```

by a single space.  For example:

```php
class Example
{
    public string $myName { get => __CLASS__; }

    public string $newName { set => ucfirst($value); }
}
```

If those two criteria are not met, then the hook block MUST be spread across multiple lines:

```php

> class Example

Reviewer 3: This example does not match the preceding paragraph, because the criteria are still met.

If multiple hooks are declared, they MUST be separated by at least a single line break.  They
MAY be separated by an additional blank line to aid readability.

If the property includes a default value, there MUST be a single space between the default value
and the opening brace.

```php
class Example
{
    public string $newName = 'Me' {
        set => ucfirst($value);
    }
}
```

If a hook is using its multi-line form, the opening brace MUST be on the same line as the hook
name, and the closing brace MUST be on its own line.  The body MUST be indented one level.

```php
class Example
{
    public string $newName = 'Me' {
        set {
            if (strlen($value) < 3) {
                throw new \Exception('Too short');
            }
            $this->newName = ucfirst($value);
        }
    }
}
```

For a `set` hook, if the argument name and type do not need to be redefined, then they MAY be omitted.

Property hooks MAY also be defined in constructor-promoted properties.  However, they
MUST be only a single hook, with a short-syntax body, defined on a single line as above.
If those criteria are not met, then the promoted property MUST NOT have any hooks defined

> inline, and SHOULD instead be defined as normal property and not promoted.

Reviewer 3: This SHOULD doesn't make sense here, because there is no other option than to use a normal property.

```php
class Example
{
    public function __construct(

>         public string $name { set => ucfirst($value; }

Reviewer 3: ```suggestion
        public string $name { set => ucfirst($value); }
```

If so, the `...` MUST NOT include any whitespace before or after. That is, the correct format is `foo(...)`.

### 4.9 Property Hooks

Object properties may also include hooks, which have a number of syntactic options.

When using the long form of hooks:

* The opening brace MUST be on the same line as the property.
* The opening brace MUST be separated from the property name or its default value by a single space.
* The closing brace MUST be on its own line, and have no comment following it.
* The entire body of the hook definition MUST be indented one level.
* The body of each hook MUST be indented one level.
* If multiple hooks are declared, they MUST be separated by at least a single line break.  They
  MAY be separated by an additional blank line to aid readability.

For example:

```php
class Example
{
    public string $newName = 'Me' {

>         set {

Reviewer 3: ```suggestion
        set(string $value) {
```

To actually show the long-long form in one of the cases.

            if (strlen($value) < 3) {
                throw new \Exception('Too short');
            }
            $this->newName = ucfirst($value);
        }
    }

    public string $department {
        get {
            return $this->values[__PROPERTY__];
        }
        set {
            $this->values[__PROPERTY__] = $value;
        }
    }
    // or
    public string $department {
        get {
            return $this->values[__PROPERTY__];
        }

        set {
            $this->values[__PROPERTY__] = $value;
        }
    }
}
```

Property hooks also support multiple short-hook variations.

For a `set` hook, if the argument name and type do not need to be redefined, then they MAY be omitted.

If a hook consists of a single expression, then PHP allows it to be shortened using `=>`.  In that case:

* There MUST be a single space on either side of the `=>` symbol.

> * The body MUST be on the same line as the hook declaration, unless it gets prohibitively long.  If it gets prohibitively long, the developer SHOULD consider not using the short-hook syntax.

Reviewer 3: This entire line is not sufficiently actionable. Combining a MUST with an unless doesn't make sense.

We already have a character limit per line, perhaps we can refer to that here?

> The body MUST be on the same line as the hook declaration and MUST NOT be wrapped. If wrapping is necessary to keep the line within the line length or desired for other reasons the short-hook syntax MUST NOT be used.

Author: Can you make an edit for that so I can merge it?

> * The body MUST be on the same line as the hook declaration, unless it gets prohibitively long.  If it gets prohibitively long, the developer SHOULD consider not using the short-hook syntax.

Reviewer 3: ```suggestion
* The body MUST be on the same line as the hook declaration and MUST NOT be wrapped. If wrapping is necessary to keep the line within the line length or desired for other reasons the short-hook syntax MUST NOT be used.
```

* The hook name MUST be separated from the opening brace and the arrow operator by a single space
* The semicolon ending of the hook MUST be separated from the closing brace by a single space.

For example:

```php
class Example
{
    public string $myName { get => __CLASS__; }

    public string $newName { set => ucfirst($value); }
}
```

Property hooks MAY also be defined in constructor-promoted properties.  However, they
MUST be only a single hook, with a short-syntax body, defined on a single line as above.
If those criteria are not met, then the promoted property MUST NOT have any hooks defined
inline.

```php
class Example
{
    public function __construct(
        public string $name { set => ucfirst($value); }
    ) {}
}
```

The following is ***not allowed*** due to the hook being too complex:

```php
class Example
{
    public function __construct(
        public string $name {
            set {
                if (strlen($value) < 3) {
                    throw new \Exception('Too short');
                }
                $this->newName = ucfirst($value);
            }
        }
    ) {}
}
```

## 4.10 Interface properties

> Abstract roperties may be defined in interfaces or abstract classes, but are required to

Reviewer: ```suggestion
Abstract properties may be defined in interfaces or abstract classes, but are required to
```

> specify if they must support `get` operations, `set` operations, or both.

Reviewer: I've opened https://github.com/php-fig/per-coding-style/pull/111 so that lowercase `must`s aren't ambiguous. I don't think this line would actually cause any confusion, but it spurred me to submit the PR.

> * The operation block MUST be on the same line as the property.

Reviewer 3: I'm not sure if I agree with the MUST here, because that effectively requires the user to use the short-form and thus is inconsistent with hooks themselves.

I find it reasonable to write the following in an interface:

```
public string $both {
    get;
    set;
}
```

And then shortening that to a single line MAY be done (consistently with hooks).

Author: There's actually a subtle difference here.  For interfaces, there can never be a body to a hook, so I see no point in the extra lines.  (If Levi ever re-proposes interface default methods that may change, but we'll cross that bridge when we get to it.)

For abstract classes, though, one could specify one hook as abstract and one as predefined.  In that case, multi-lining is probably better.

We should probably address those separately.

Reviewer 3: > so I see no point in the extra lines

The point is _allowing_ for an uniform formatting of similar things. In this case: A property declaration with a list of “operations”.

I also find it much easier to scan the code vertically rather than horizontally, because the of the established pattern of “a line-break indicates that something new begins, with indentation indicating the relationship to the previous line”.

* If multiple hooks are declared, they MUST be separated by at least a single line break.  They
  MAY be separated by an additional blank line to aid readability.

For example:

```php
class Example
{
    public string $newName = 'Me' {
        set(string $value) {
            if (strlen($value) < 3) {
                throw new \Exception('Too short');
            }
            $this->newName = ucfirst($value);
        }
    }

    public string $department {
        get {
            return $this->values[__PROPERTY__];
        }
        set {
            $this->values[__PROPERTY__] = $value;
        }
    }
    // or
    public string $department {
        get {
            return $this->values[__PROPERTY__];
        }

        set {
            $this->values[__PROPERTY__] = $value;
        }
    }
}
```

Property hooks also support multiple short-hook variations.

For a `set` hook, if the argument name and type do not need to be redefined, then they MAY be omitted.

If a hook consists of a single expression, then PHP allows it to be shortened using `=>`.  In that case:

* There MUST be a single space on either side of the `=>` symbol.

> * The body MUST be on the same line as the hook declaration and MUST NOT be wrapped. If wrapping is necessary to keep the line within the line length or desired for other reasons the short-hook syntax MUST NOT be used.

Author: Having spent more time working with hooks myself now, I am not sure I agree with this line anymore.  Ternaries are perfectly reasonable as a single-expression hook body, and those are often multilined for clarity.  I can certainly see match statements making sense in some cases, and those are obviously multi-line.  

I think I want to soften this to "must start on the same line", and then let other rules around how to wrap things apply.  Comments?

Reviewer 4: I agree. It's a similar situation to short closures, which also sometimes have an expression which it makes sense to split across multiple lines.

Author: Absent any feedback, I went ahead and made this change.
